packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -0,0 +1,406 @@
+/// arbitrary child widget.
+///
+/// Use [InputDecorator] to create widgets that look and behave like the [Input]
+/// widget.
+///
+/// Requires one of its ancestors to be a [Material] widget.
+///
+/// See also:
+///
+/// * [Input], which combines an [InputDecorator] with an [InputField].
+class InputDecorator extends StatefulWidget {
+  InputDecorator({
+    Key key,
+    @required this.decoration,
+    this.baseStyle,
+    this.isFocused: false,
+    this.isEmpty: false,
+    this.child,
+  }) : super(key: key);
+
+  final InputDecoration decoration;
+
+  final TextStyle baseStyle;
+
+  /// True if the hint and label should be displayed as if the child had the focus.
+  ///
+  /// Defaults to false.
+  final bool isFocused;
+
+  /// Should the hint and label be displayed as if no value had been input
+  /// to the child.
+  ///
+  /// Defaults to false.
+  final bool isEmpty;
+
+  /// The widget below this widget in the tree.
+  final Widget child;
+
+  @override
+  _InputDecoratorState createState() => new _InputDecoratorState();
+
+  @override
+  void debugFillDescription(List<String> description) {
+    super.debugFillDescription(description);
+    description.add('decoration: $decoration');
+    description.add('baseStyle: $baseStyle');
+    description.add('isFocused: $isFocused');
+    description.add('isEmpty: $isEmpty');
+  }
+}
+
+const Duration _kTransitionDuration = const Duration(milliseconds: 200);
+const Curve _kTransitionCurve = Curves.fastOutSlowIn;
+
+class _InputDecoratorState extends State<InputDecorator> {
+
+  Color _getActiveColor(ThemeData themeData) {
+    if (config.isFocused) {
+      switch (themeData.brightness) {
+        case Brightness.dark:
+          return themeData.accentColor;
+        case Brightness.light:
+          return themeData.primaryColor;
+      }
+    }
+    return themeData.hintColor;
+  }
+
+  Widget _buildContent(Color borderColor, double topPadding, bool isDense) {
+    final double bottomPadding = isDense ? 8.0 : 1.0;
+    const double bottomBorder = 2.0;
+    final double bottomHeight = isDense ? 14.0 : 18.0;
+
+    final EdgeInsets padding = new EdgeInsets.only(top: topPadding, bottom: bottomPadding);
+    final EdgeInsets margin = new EdgeInsets.only(bottom: bottomHeight - (bottomPadding + bottomBorder));
+
+    if (config.decoration.hideDivider) {
+      return new Container(
+        margin: margin + new EdgeInsets.only(bottom: bottomBorder),
+        padding: padding,
+        child: config.child,
+      );
+    }
+
+    return new AnimatedContainer(
+      margin: margin,
+      padding: padding,
+      duration: _kTransitionDuration,
+      curve: _kTransitionCurve,
+      decoration: new BoxDecoration(
+        border: new Border(
+          bottom: new BorderSide(
+            color: borderColor,
+            width: bottomBorder,
+          ),
+        ),
+      ),
+      child: config.child,
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
```

Comment:
if we factored this out into the InputDecoration class itself, then you could pass in entirely novel decorations, much as we do with DecoratedBox.

Actually that kind of seems like a primitive we should have. `build(BuildContext context, Widget child)`. AnimatedBuilder does something like that, though it passes in a Listenable as well IIRC.